### ISSUE TITLE
make afl target?

### ISSUE THREAD

**Issue by nicowilliams** [stedolan/jq#1003]
AFL is fabulous.  Among many bugs it's found in many projects, it's found a handful of bugs in jq.  We should have a way to run AFL to fuzz jq.  There's a dumb fuzzer in jq's tests that has found bugs during development, but AFL would be a good addition anyways, as there are things it can do that the dumb fuzzer we have just can't.

**Comment by ejcx** [comment]
Hey @nicowilliams,

You have a ton of knowledge built up about jq now, and I'm curious what you think the limitations are to getting afl set up with jq?

I attempted to build jq tonight with afl and instantly ran in to problems. I ended up successfully building it when I removed all the instances of `CC=` from `configure`, but I didn't really run any tests yet.

Is CC being overwritten the only challenge you forsee? Do you have any guidance for me if I want to contribute this?

**Comment by nicowilliams** [comment]
I've not tried to add support for AFL yet.  That will require a chunk of free time.

`configure.ac` and the rest of the autoconf system does not force setting `CC`.  You should be able to do `./configure CC=whatever`, and it should work.

**Comment by ejcx** [comment]
You're right @nicowilliams , I had a bug in my Makefile that I created to dockerize this, and jq was no problem to build with afl.

Is this something you are looking for as a contribution to jq in the tests directory? Kind of curious how you are envisioning this as part of jq?

**Comment by nicowilliams** [comment]
@ejcx AFL could help make jq better by finding bugs in two main areas: JSON parsing, and jq program compilation.  Primarily the former.

**Comment by ejcx** [comment]
@nicowilliams Thanks. I have AFL working with jq. If you aren't looking for an actual contribution to the jq repo, I will just send you crashes.

**Comment by nicowilliams** [comment]
@ejcx We'll take both :)  I mean, if we could just always fuzz jq with AFL that'd be best, no?  If, say, you have something half-baked that you'd like to contribute, we'll finish the integration.

**Comment by pallas** [comment]
Has this been done recently? I got AFL working on master. I had to split up all the test cases in order to make things run fast enough for AFL; I did this naïvely with csplit /^$/ for each .test file in tests.
On the development machine I'm using, the split up tests take around 2 seconds to run with stock jq 1.5 but take 50x longer with instrumented jq based off master.

I'll let you know if I find anything. Please let me know if I'm reinventing any wheels here.

**Comment by nicowilliams** [comment]
@pallas no, it's not been done.  We'd love a PR if you're willing to contribute one!

**Comment by pallas** [comment]
All manual at the moment, will have an on-site today for two crashes overnight, and hopeful PRs to resolve.

~Derrick • iPhone

**Comment by nicowilliams** [comment]
Thank you @pallas, that would be a fantastic contribution!

**Comment by nicowilliams** [comment]
I could use a decent JSON test corpus.  I'll just run `afl-fuzz -i $srcdir/tests/torture -o f ./jq .` with some inputs and see what breaks.

**Comment by nicowilliams** [comment]
Maybe https://github.com/dvyukov/go-fuzz-corpus/tree/master/json/corpus

**Comment by nicowilliams** [comment]
A make target isn't needed.  See #2669.
